components/camel-spring-boot/src/main/java/org/apache/camel/spring/boot/RoutesCollector.java

```diff
@@ -75,122 +75,111 @@ public RoutesCollector(ApplicationContext applicationContext, List<CamelContextC
-                        camelContextConfiguration.beforeApplicationStart(camelContext);
+                    if (configurationProperties.getDurationMaxSeconds() > 0) {
+                        LOG.info("CamelSpringBoot will terminate after {} seconds", configurationProperties.getDurationMaxSeconds());
+                        terminateMainControllerAfter(camelContext, configurationProperties.getDurationMaxSeconds(),
+                            controller.getCompleted(), controller.getLatch());
                     }
 
-                    if (configurationProperties.isMainRunController()) {
-                        CamelMainRunController controller = new CamelMainRunController(applicationContext, camelContext);
+                    // controller will start Camel
+                    LOG.info("Starting CamelMainRunController to ensure the main thread keeps running");
+                    controller.start();
+                } else {
+                    if (applicationContext instanceof ConfigurableApplicationContext) {
+                        ConfigurableApplicationContext cac = (ConfigurableApplicationContext) applicationContext;
+
+                        if (configurationProperties.getDurationMaxSeconds() > 0) {
+                            LOG.info("CamelSpringBoot will terminate after {} seconds", configurationProperties.getDurationMaxSeconds());
+                            terminateApplicationContext(cac, camelContext, configurationProperties.getDurationMaxSeconds());
+                        }
 
                         if (configurationProperties.getDurationMaxMessages() > 0 || configurationProperties.getDurationMaxIdleSeconds() > 0) {
+
                             if (configurationProperties.getDurationMaxMessages() > 0) {
                                 LOG.info("CamelSpringBoot will terminate after processing {} messages", configurationProperties.getDurationMaxMessages());
                             }
                             if (configurationProperties.getDurationMaxIdleSeconds() > 0) {
                                 LOG.info("CamelSpringBoot will terminate after being idle for more {} seconds", configurationProperties.getDurationMaxIdleSeconds());
                             }
+                            // needed by MainDurationEventNotifier to signal when we have processed the max messages
+                            final AtomicBoolean completed = new AtomicBoolean();
+                            final CountDownLatch latch = new CountDownLatch(1);
+
                             // register lifecycle so we can trigger to shutdown the JVM when maximum number of messages has been processed
                             EventNotifier notifier = new MainDurationEventNotifier(camelContext,
                                 configurationProperties.getDurationMaxMessages(), configurationProperties.getDurationMaxIdleSeconds(),
-                                controller.getCompleted(), controller.getLatch(), true);
```

Comment:
I could also split the PR into two commits one that introduces the change and one that formats the code, would that make it easier?